### PR TITLE
Implement `Array.__iter__` [test-upstream]

### PR DESCRIPTION
**Pull request by jsignell**
ref https://github.com/pandas-dev/pandas/issues/38645

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

See the linked pandas issue for reference. This PR adds an `__iter__` method that greedily computes the array. I am unsure if this behavior would be too surprising. For additional context, the `dd.Series.__iter__` method already computes the data. 

Here's a comparison between this PR and main:

## This PR
```python
In [1]: import dask.array as da
   ...: 
   ...: a = da.ones((12,), chunks=4)
   ...: list(a)
Out[1]: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]

In [2]: 1 in a
Out[2]: True

In [3]: [value for value in a]
Out[3]: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]

In [4]: [a[i] for i in range(len(a))]
Out[4]: 
[dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>]
```

## Main
```python
In [1]: import dask.array as da
   ...: 
   ...: a = da.ones((12,), chunks=4)
   ...: list(a)
Out[1]: 
[dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>]

In [2]: 1 in a
Out[2]: True

In [3]: [value for value in a]
Out[3]: 
[dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>]

In [4]: [a[i] for i in range(len(a))]
Out[4]: 
[dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>,
 dask.array<getitem, shape=(), dtype=float64, chunksize=(), chunktype=numpy.ndarray>]
```

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -1341,6 +1341,13 @@ def chunks(self, chunks):
             "  x.compute_chunk_sizes()"
         )
 
+    def __iter__(self):
+        """Iterating over a dask array forces values to be evaluated"""
+        for block in self.blocks:
+            array = block.compute()
+            for row in array:
+                yield row
```
Comment (mrocklin) [inline dask/array/core.py:1349]: What happens if the array is 2d?

**Review by mrocklin — COMMENTED** [review]
Shouldn't there be some other protocol than iter that gets triggered first?  What about one of the array protocols?

File: dask/array/core.py (hunk 1 above)
Comment (jsignell) [reply dask/array/core.py:1349]:
```python
In [1]: import dask.array as da
   ...: 
   ...: a = da.ones((12, 3), chunks=(4,1))
   ...: list(a)
Out[1]: 
[array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.]),
 array([1., 1., 1.])]
```

**Comment by jsignell** [comment]
> Shouldn't there be some other protocol than iter that gets triggered first? What about one of the array protocols?

That does seem sensible :)

**Comment by jsignell** [comment]
It seems like this can be fixed on the pandas side, so I'll go ahead and close this, but I do think we should be doing something intentional with `__iter__`. I'm just not sure what is least surprising :)

**Pull request closed without merging**